keps/draft-20180806-namespace-initializer.md

```diff
@@ -0,0 +1,394 @@
+- V2 Namespace
+
+## Proposal
+
+### User Stories
+
+#### Automatic Policy Population
+
+In a large enterprise, a single Kubernetes cluster is shared by multiple engineering teams.
+The cluster administrators allow the engineering teams to do self-service namespace creation (using `kubectl create namespace`) to make the process simpler,
+but they still want to enforce the policies in each namespace automatically, including RBAC, ResourceQuota, NetworkPolicy etc.
+They define these policies in a centralized places (e.g. git, corp AD, etc.),
+and use a controller watching namespace creation and populate the predefined policies during namespace creation before the creator gains full permission inside the namespace.
+
+This is the user story demonstrated by the [Namespace Population](https://github.com/kubernetes/community/pull/2177) proposal.
+
+#### Automated External Service Provisioning
+
+The Kubernetes cluster administrator wants to associate namespaces with log analysis system backed by ElasticSearch.
+During the namespace creation, they want to automatically provision ElasticSearch with a new index dedicated for logs sent from this namespace.
+They deploy a controller to perform the provisioning when watch a new namespace creation,
+and they don't want the users to access the namespace before the provisioning is completed.
+
+### Implementation Details
+
+The mechanism is derived from exising [Initializers](https://kubernetes.io/docs/reference/access-authn-authz/extensible-admission-controllers/#initializers).
+For more background and initial discussion, please read comments in [Namespace Population KEP](https://github.com/kubernetes/community/pull/2177).
+
+Briefly, namespace initialization involves a few components:
+
+- A built-in admission controller `NamespaceInitializers`
+- A resource of type `NamespaceInitializerConfiguration`
+- Some logic in storage layer to keep the backward-compatible behavior of the API
+- A few namespace initialization controllers: custom controllers which initialize a namespace by creating resources inside the namespace
+
+They work in the process of:
+
+- The namespace initialization controllers register themselves by creating `NamespaceInitializerConfiguration` resources
+- When a namespace is newly created
+  - The `NamespaceInitializers` admission controller aggregates all `NamespaceInitializersConfiguration` resources and generate `spec.Initializers` in this namespace resource
+  - Once the namespace resource is persisted, the following happens in parallel:
+    - Namespace initialization controllers get notified about the newly created namespace,
+      and they start creating resources inside namespace and also update `spec.Initializers` to remove themselves once the initialization is done
+    - The storage layer will wait until `spec.Initializers` becomes empty before returning the response back to client, unless `includeUninitialized` is true in the request.
```

Comment:
I understand holding the creation request is complex and using `includeUninitialized` can be problematic. However, this seems to be the only way to ensure _behavioral compatibility_ with existing client code - which assumes the readiness of a resource upon receiving the creation response. That's also a reason to introduce status condition in namespace resource, with which we can gradually improve the client code (at least client-go and kubectl) to determine the readiness of a resource based on status condition.

For example:

```yaml
apiVersion: v1
kind: Namespace
metadata:
  name: ns1
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: ns1
  name: app1
...
```

There are a lot of usage of above manifest with a single `kubectl apply -f`. It will break without holding the namespace creation request if namespace is not ready because of being initialized by some controller while the creation response reaches kubectl. For example, in self-service namespace creation, the creator may only have the permission to `create` a namespace, and waits until `NamespaceTemplate` to populate `admin` permission inside the namespace. The single `kubectl apply` will break without initialization mechanism.

I think a reasonable plan will be still including the mechanism of `includeUninitialized` flag to make sure nothing breaks. And later deprecate this complex logic when most of the clients are patched and working with status condition (this will take significant long time, probably a few years).